DigitalLearningSolutions.Data/DataServices/RegistrationDataService.cs

```diff
@@ -20,39 +21,146 @@ public RegistrationDataService(IDbConnection connection)
             this.connection = connection;
         }
 
-        public string RegisterDelegate(DelegateRegistrationModel delegateRegistrationModel)
+        public string RegisterNewUserAndDelegateAccount(DelegateRegistrationModel delegateRegistrationModel)
         {
-            var values = new
+            var userValues = new
+                {
+                    delegateRegistrationModel.FirstName,
+                    delegateRegistrationModel.LastName,
+                    delegateRegistrationModel.PrimaryEmail,
+                    delegateRegistrationModel.JobGroup,
+                    delegateRegistrationModel.Active,
+                    PasswordHash = "temp",
+                    ProfessionalRegistrationNumber = (string?)null,
+                };
+
+            var userId = connection.QuerySingle<int>(
+                @"INSERT INTO Users
+                    (
+                        PrimaryEmail,
+                        PasswordHash,
+                        FirstName,
+                        LastName,
+                        JobGroupID,
+                        ProfessionalRegistrationNumber,
+                        Active
+                    )
+                    OUTPUT Inserted.ID
+                    VALUES
+                    (
+                        @primaryEmail,
+                        @passwordHash,
+                        @firstName,
+                        @lastName,
+                        @jobGroup,
+                        @professionalRegistrationNumber,
+                        @active
+                    )",
+                    userValues
+                );
+
+            var initials = delegateRegistrationModel.FirstName.Substring(1) +
+                           delegateRegistrationModel.LastName.Substring(1);
+
+            // this SQL is reproduced mostly verbatim from the uspSaveNewCandidate_V10 procedure in the legacy codebase.
+            var candidateNumber = connection.QueryFirst<string>(

```

Comment:
Can we put this inside a transaction to ensure HOLDLOCK gives us a lock that lasts until the end of the query? May not be necessary (I'm not exactly sure of the internals of dapper or SQL Server defaults) but no harm in it.
In order to wrap it all in a db transaction, I think we need to pass the transaction scope to the dapper query call explicitly, which we don't often do elsewhere (See https://dapper-tutorial.net/transaction). Normally I think we just have a C# transaction that rolls back if Complete() isn't called, which may include multiple db transactions (see https://docs.microsoft.com/en-us/dotnet/api/system.transactions.transactionscope?view=net-6.0#examples)

Reply:
My interpretation was just the candidate number bit - I may have misunderstood @SteveJacksonSoft 's meaning.